File: ramps/golang/scenarios/chat_sse.go
package scenarios

import (
	"bufio"
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/csv"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math/rand"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync/atomic"
	"time"
)

// DefaultQueries danh sách câu hỏi kiểm thử mẫu chung cho chatbot (dự phòng khi không cấu hình .env).
var DefaultQueries = []string{}

// LoadDotEnv tìm và nạp các biến môi trường từ file .env vào process (os.Setenv) nếu chưa tồn tại.
func LoadDotEnv(path string) map[string]string {
	var candidates []string
	if path != "" {
		candidates = append(candidates, path)
		if !filepath.IsAbs(path) {
			candidates = append(candidates,
				filepath.Join(".", path),
				filepath.Join("..", path),
				filepath.Join("..", "..", path),
			)
		}
	} else {
		candidates = append(candidates,
			".env",
			filepath.Join("..", ".env"),
			filepath.Join("..", "..", ".env"),
		)
	}

	var targetPath string
	for _, c := range candidates {
		if fi, err := os.Stat(c); err == nil && !fi.IsDir() {
			targetPath = c
			break
		}
	}
	if targetPath == "" {
		return nil
	}

	content, err := os.ReadFile(targetPath)
	if err != nil {
		return nil
	}

	envVars := make(map[string]string)
	lines := strings.Split(string(content), "\n")
	i := 0
	for i < len(lines) {
		line := strings.TrimSpace(lines[i])
		i++
		if line == "" || strings.HasPrefix(line, "#") || !strings.Contains(line, "=") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		k := strings.TrimSpace(parts[0])
		v := strings.TrimSpace(parts[1])

		// Xử lý chuỗi multiline nằm trong dấu ngoặc kép hoặc đơn
		if (strings.HasPrefix(v, "\"") && !(strings.HasSuffix(v, "\"") && len(v) > 1)) ||
			(strings.HasPrefix(v, "'") && !(strings.HasSuffix(v, "'") && len(v) > 1)) {
			quoteChar := string(v[0])
			valParts := []string{v[1:]}
			for i < len(lines) {
				nextLine := lines[i]
				i++
				if strings.HasSuffix(strings.TrimRight(nextLine, "\r"), quoteChar) {
					trimmed := strings.TrimRight(nextLine, "\r")
					valParts = append(valParts, trimmed[:len(trimmed)-1])
					break
				}
				valParts = append(valParts, strings.TrimRight(nextLine, "\r"))
			}
			v = strings.Join(valParts, "\n")
		} else if (strings.HasPrefix(v, "\"") && strings.HasSuffix(v, "\"")) ||
			(strings.HasPrefix(v, "'") && strings.HasSuffix(v, "'")) {
			v = v[1 : len(v)-1]
		}

		v = strings.ReplaceAll(v, "\\n", "\n")
		envVars[k] = v
		if os.Getenv(k) == "" {
			_ = os.Setenv(k, v)
		}
	}
	return envVars
}

// ParseQueries phân tích chuỗi câu hỏi từ .env (hỗ trợ JSON array, xuống dòng, | hoặc ;).
func ParseQueries(raw string) []string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return nil
	}

	// 1. JSON Array
	if strings.HasPrefix(raw, "[") && strings.HasSuffix(raw, "]") {
		var list []string
		if err := json.Unmarshal([]byte(raw), &list); err == nil && len(list) > 0 {
			var res []string
			for _, item := range list {
				item = strings.TrimSpace(item)
				if item != "" {
					res = append(res, item)
				}
			}
			if len(res) > 0 {
				return res
			}
		}
	}

	// 2. Phân tách theo dòng (\n)
	if strings.Contains(raw, "\n") {
		var res []string
		for _, line := range strings.Split(raw, "\n") {
			line = strings.TrimSpace(line)
			if line != "" {
				res = append(res, line)
			}
		}
		if len(res) > 0 {
			return res
		}
	}

	// 3. Phân tách theo dấu pipe (|)
	if strings.Contains(raw, "|") {
		var res []string
		for _, item := range strings.Split(raw, "|") {
			item = strings.TrimSpace(item)
			if item != "" {
				res = append(res, item)
			}
		}
		if len(res) > 0 {
			return res
		}
	}

	// 4. Phân tách theo dấu chấm phẩy (;)
	if strings.Contains(raw, ";") {
		var res []string
		for _, item := range strings.Split(raw, ";") {
			item = strings.TrimSpace(item)
			if item != "" {
				res = append(res, item)
			}
		}
		if len(res) > 0 {
			return res
		}
	}

	return []string{raw}
}

// ResolveQueries xác định danh sách câu hỏi kiểm thử theo độ ưu tiên:
// 1. Cờ dòng lệnh --queries (file txt/csv)
// 2. Biến CHAT_QUERIES_FILE hoặc QUERIES_FILE từ .env
// 3. Biến CHAT_QUERIES hoặc QUERIES từ .env
// 4. Danh sách mặc định DefaultQueries
func ResolveQueries(queriesFile string, envFile string) []string {
	LoadDotEnv(envFile)

	// 1. Cờ dòng lệnh --queries
	if queriesFile != "" {
		content, err := os.ReadFile(queriesFile)
		if err == nil {
			var lines []string
			for _, line := range strings.Split(string(content), "\n") {
				line = strings.TrimSpace(line)
				if line != "" {
					lines = append(lines, line)
				}
			}
			if len(lines) > 0 {
				fmt.Printf("queries  : %d câu hỏi nạp từ cờ --queries (%s)\n", len(lines), filepath.Base(queriesFile))
				return lines
			}
		}
		fmt.Fprintf(os.Stderr, "CẢNH BÁO: không đọc được file --queries: %s\n", queriesFile)
	}

	// 2. Biến file từ .env
	qFile := os.Getenv("CHAT_QUERIES_FILE")
	if qFile == "" {
		qFile = os.Getenv("QUERIES_FILE")
	}
	if qFile != "" {
		candidates := []string{
			qFile,
			filepath.Join("..", qFile),
			filepath.Join("..", "..", qFile),
		}
		for _, c := range candidates {
			content, err := os.ReadFile(c)
			if err == nil {
				var lines []string
				for _, line := range strings.Split(string(content), "\n") {
					line = strings.TrimSpace(line)
					if line != "" {
						lines = append(lines, line)
					}
				}
				if len(lines) > 0 {
					fmt.Printf("queries  : %d câu hỏi nạp từ file %s (qua .env)\n", len(lines), filepath.Base(c))
					return lines
				}
			}
		}
	}

	// 3. Chuỗi danh sách câu hỏi trong CHAT_QUERIES hoặc QUERIES (.env)
	rawEnv := os.Getenv("CHAT_QUERIES")
	if rawEnv == "" {
		rawEnv = os.Getenv("QUERIES")
	}
	if rawEnv != "" {
		parsed := ParseQueries(rawEnv)
		if len(parsed) > 0 {
			fmt.Printf("queries  : %d câu hỏi nạp từ .env (CHAT_QUERIES)\n", len(parsed))
			return parsed
		}
	}

	// 4. Dự phòng mặc định
	fmt.Println("queries  : dùng danh sách câu hỏi mẫu mặc định (chưa cấu hình CHAT_QUERIES trong .env)")
	return DefaultQueries
}

// MintJWT sinh JWT HS256 nội bộ với role super_admin.
func MintJWT(secret string, ttlSeconds int64) (string, error) {
	headerJSON := `{"alg":"HS256","typ":"JWT"}`
	headerEnc := base64.RawURLEncoding.EncodeToString([]byte(headerJSON))

	payload := map[string]any{
		"roles": []string{"super_admin"},
		"exp":   time.Now().Unix() + ttlSeconds,
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	payloadEnc := base64.RawURLEncoding.EncodeToString(payloadBytes)

	signingInput := headerEnc + "." + payloadEnc
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(signingInput))
	sigEnc := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))

	return signingInput + "." + sigEnc, nil
}

type User struct {
	ID    string
	Token string
}

type UserPool struct {
	users []User
	idx   uint64
}

func NewUserPool(users []User) *UserPool {
	return &UserPool{users: users}
}

func (p *UserPool) Len() int {
	return len(p.users)
}

func (p *UserPool) Next() (string, string) {
	if len(p.users) == 0 {
		return "", ""
	}
	i := atomic.AddUint64(&p.idx, 1) - 1
	u := p.users[i%uint64(len(p.users))]
	return u.ID, u.Token
}

// LoadUsers nạp danh sách user từ file CSV, nhận diện linh hoạt các cột token và user_id.
func LoadUsers(csvPath string) ([]User, error) {
	f, err := os.Open(csvPath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(records) == 0 {
		return nil, fmt.Errorf("%s: file rỗng", csvPath)
	}

	header := records[0]
	tokenCol := -1
	userCol := -1
	for i, h := range header {
		clean := strings.ToLower(strings.TrimSpace(h))
		if clean == "token" || clean == "user_token" || clean == "access_token" || clean == "api_key" {
			if tokenCol == -1 {
				tokenCol = i
			}
		}
		if clean == "user_id" || clean == "userid" || clean == "id" || clean == "user" {
			if userCol == -1 {
				userCol = i
			}
		}
	}

	// Lượt 2: Tìm kiếm gần đúng (fuzzy match) nếu chưa tìm thấy khớp chính xác.
	// Nhận diện bất kỳ cột nào chứa chuỗi "token" (không phân biệt hoa/thường, ví dụ: "tokenUserChatbot", "auth_token",...).
	// Cơ chế này giúp tự động tương thích với các định dạng CSV từ nhiều nguồn, tránh việc LoadUsers báo lỗi
	// khiến hệ thống tự động rơi về chế độ dùng 1 token dùng chung làm nghẽn rate-limit của tài khoản.
	if tokenCol == -1 {
		for i, h := range header {
			if strings.Contains(strings.ToLower(strings.TrimSpace(h)), "token") {
				tokenCol = i
				break
			}
		}
	}

	if tokenCol == -1 {
		return nil, fmt.Errorf("%s: không tìm thấy cột nào chứa token trong header %v", csvPath, header)
	}

	var users []User
	for i, row := range records[1:] {
		if tokenCol >= len(row) {
			continue
		}
		tok := strings.TrimSpace(row[tokenCol])
		if tok == "" {
			continue
		}
		uid := fmt.Sprintf("%d", i+1)
		if userCol != -1 && userCol < len(row) {
			val := strings.TrimSpace(row[userCol])
			if val != "" {
				uid = val
			}
		}
		users = append(users, User{ID: uid, Token: tok})
	}
	if len(users) == 0 {
		return nil, fmt.Errorf("%s: không có dòng nào hợp lệ", csvPath)
	}
	return users, nil
}

// ChatSSEScenario kịch bản kiểm thử API Chatbot streaming SSE.
type ChatSSEScenario struct {
	url       string
	headers   map[string]string
	timeout   time.Duration
	pool      *UserPool
	queries   []string
	cacheBust bool
}

func NewChatSSEScenario(
	url string,
	headers map[string]string,
	timeout time.Duration,
	pool *UserPool,
	queries []string,
	cacheBust bool,
) *ChatSSEScenario {
	hdrs := make(map[string]string)
	for k, v := range headers {
		hdrs[k] = v
	}
	if _, ok := hdrs["Content-Type"]; !ok {
		hdrs["Content-Type"] = "application/json"
	}
	if len(queries) == 0 {
		queries = DefaultQueries
	}
	return &ChatSSEScenario{
		url:       url,
		headers:   hdrs,
		timeout:   timeout,
		pool:      pool,
		queries:   queries,
		cacheBust: cacheBust,
	}
}

func (s *ChatSSEScenario) Name() string        { return "chat-sse" }
func (s *ChatSSEScenario) MetricLabel() string { return "ttft" }
func (s *ChatSSEScenario) Setup(ctx context.Context, client *http.Client) error {
	return nil
}

func (s *ChatSSEScenario) Execute(ctx context.Context, client *http.Client) Record {
	q := s.queries[rand.Intn(len(s.queries))]
	if s.cacheBust {
		q = fmt.Sprintf("%s (ma tham chieu %d)", q, rand.Intn(900000)+100000)
	}
	uid, token := s.pool.Next()

	payload := map[string]any{
		"user_token": token,
		"summary":    "",
		"history": []map[string]string{
			{"role": "user", "text": "Xin chào"},
			{"role": "assistant", "text": "Chào bạn, mình là chatbot. Bạn cần hỗ trợ gì?"},
			{"role": "user", "text": q},
		},
	}
	bodyBytes, _ := json.Marshal(payload)

	tStart := float64(time.Now().UnixNano()) / 1e9
	t0 := time.Now()

	endSeen := false
	rec := Record{
		Type:    "req",
		TStart:  tStart,
		UserID:  uid,
		EndSeen: &endSeen,
		Ok:      false,
	}

	reqCtx, cancel := context.WithTimeout(ctx, s.timeout)
	defer cancel()

	httpReq, err := http.NewRequestWithContext(reqCtx, "POST", s.url, bytes.NewReader(bodyBytes))
	if err != nil {
		rec.Err = strPtr(err.Error())
		elapsed := float64(time.Since(t0).Microseconds()) / 1000.0
		rec.TotalMs = &elapsed
		return rec
	}

	for k, v := range s.headers {
		httpReq.Header.Set(k, v)
	}

	resp, err := client.Do(httpReq)
	if err != nil {
		if errors.Is(err, context.Canceled) {
			rec.Err = strPtr("abandoned")
		} else {
			rec.Err = strPtr(err.Error())
		}
		elapsed := float64(time.Since(t0).Microseconds()) / 1000.0
		rec.TotalMs = &elapsed
		return rec
	}
	defer resp.Body.Close()

	rec.Status = &resp.StatusCode
	if resp.StatusCode != http.StatusOK {
		_, _ = io.Copy(io.Discard, resp.Body)
		rec.Err = strPtr(fmt.Sprintf("http_%d", resp.StatusCode))
		elapsed := float64(time.Since(t0).Microseconds()) / 1000.0
		rec.TotalMs = &elapsed
		return rec
	}

	reader := bufio.NewReader(resp.Body)
	var event string
	for {
		line, err := reader.ReadString('\n')
		if line != "" {
			trimmed := strings.TrimRight(line, "\r\n")
			if strings.HasPrefix(trimmed, "event:") {
				event = strings.TrimSpace(trimmed[6:])
			} else if strings.HasPrefix(trimmed, "data:") {
				now := float64(time.Since(t0).Microseconds()) / 1000.0
				if event == "message" && rec.TtftMs == nil {
					rec.TtftMs = &now
				} else if event == "notice" && rec.NoticeMs == nil {
					rec.NoticeMs = &now
				} else if event == "message_end" {
					*rec.EndSeen = true
				} else if event == "error" {
					rec.Err = strPtr("sse_error")
				}
			}
		}
		if err != nil {
			if errors.Is(err, context.Canceled) {
				// Request bị hủy bởi step runner khi hết thời gian settle cuối bậc kiểm thử.
				// Đánh dấu err = "abandoned" để đồng bộ với bản Python; report.py nhận diện chính xác
				// chuỗi lỗi này làm một trong các chỉ số phát hiện điểm gãy quá tải (knee condition).
				if rec.Err == nil {
					rec.Err = strPtr("abandoned")
				}
			} else if err != io.EOF {
				if rec.Err == nil {
					rec.Err = strPtr(err.Error())
				}
			}
			break
		}
	}

	elapsed := float64(time.Since(t0).Microseconds()) / 1000.0
	rec.TotalMs = &elapsed
	rec.Ok = *rec.EndSeen && rec.Err == nil
	return rec
}

func strPtr(s string) *string {
	return &s
}

File: ramps/python/scenarios/chat_sse.py
from __future__ import annotations

import argparse
import csv
import os
import random
import sys
import time
from pathlib import Path
from typing import Any

import httpx

from .base import Scenario

def find_default_users_csv() -> Path:
    """Tìm đường dẫn mặc định đến users.csv (ưu tiên thư mục hiện tại, rồi thư mục gốc repo)."""
    candidates = [
        Path("users.csv"),
        Path.cwd() / "users.csv",
        Path(__file__).resolve().parents[2] / "users.csv",
        Path(__file__).resolve().parents[3] / "users.csv",
        Path(__file__).resolve().parent.parent / "users.csv",
    ]
    return next((c for c in candidates if c.is_file()), Path("users.csv"))


# Đường dẫn mặc định đến file CSV chứa thông tin người dùng và token (nằm ở thư mục gốc của project)
DEFAULT_USERS_CSV = find_default_users_csv()

# Danh sách câu hỏi kiểm thử mẫu chung cho chatbot (dự phòng khi không cấu hình .env)
DEFAULT_QUERIES = [
    "Xin chào, bạn có thể hỗ trợ những tác vụ gì?",
    "Hãy tóm tắt ngắn gọn các điểm chính của tài liệu này",
    "Hướng dẫn tôi cách tích hợp API vào hệ thống",
    "Giải thích giúp tôi nguyên lý hoạt động của kiến trúc microservices",
    "Gợi ý cho tôi một số giải pháp tối ưu hóa hiệu năng",
    "Chào bạn",
]


def load_dotenv(path: str | Path | None = None) -> dict[str, str]:
    """Tìm và nạp các biến môi trường từ file .env vào os.environ nếu chưa tồn tại."""
    candidates: list[Path] = []
    if path:
        p = Path(path)
        candidates.append(p)
        if not p.is_absolute():
            candidates.extend([
                Path.cwd() / p,
                Path(__file__).resolve().parent.parent / p,
                Path(__file__).resolve().parents[2] / p,
            ])
    else:
        candidates.extend([
            Path(".env"),
            Path.cwd() / ".env",
            Path(__file__).resolve().parent.parent / ".env",
            Path(__file__).resolve().parents[2] / ".env",
        ])

    target = next((c for c in candidates if c.is_file()), None)
    if not target:
        return {}

    env_vars: dict[str, str] = {}
    content = target.read_text("utf-8")
    lines = content.splitlines()
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        i += 1
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        k = k.strip()
        v = v.strip()

        # Xử lý chuỗi multiline nằm trong dấu nháy kép hoặc đơn
        if (v.startswith('"') and not (v.endswith('"') and len(v) > 1)) or (
            v.startswith("'") and not (v.endswith("'") and len(v) > 1)
        ):
            quote_char = v[0]
            val_parts = [v[1:]]
            while i < len(lines):
                next_line = lines[i]
                i += 1
                if next_line.rstrip().endswith(quote_char):
                    val_parts.append(next_line.rstrip()[:-1])
                    break
                val_parts.append(next_line)
            v = "\n".join(val_parts)
        elif (v.startswith('"') and v.endswith('"')) or (v.startswith("'") and v.endswith("'")):
            v = v[1:-1]

        v = v.replace("\\n", "\n")
        env_vars[k] = v
        if k not in os.environ:
            os.environ[k] = v

    return env_vars


def parse_queries(raw: str) -> list[str]:
    """Phân tích chuỗi câu hỏi từ .env thành danh sách (hỗ trợ JSON array, xuống dòng, | hoặc ;)."""
    raw = raw.strip()
    if not raw:
        return []
    # 1. JSON array
    if raw.startswith("[") and raw.endswith("]"):
        try:
            import json

            data = json.loads(raw)
            if isinstance(data, list):
                res = [str(x).strip() for x in data if str(x).strip()]
                if res:
                    return res
        except Exception:
            pass
    # 2. Xuống dòng
    if "\n" in raw:
        lines = [ln.strip() for ln in raw.split("\n") if ln.strip()]
        if lines:
            return lines
    # 3. Ký tự phân cách pipe (|)
    if "|" in raw:
        lines = [ln.strip() for ln in raw.split("|") if ln.strip()]
        if lines:
            return lines
    # 4. Ký tự phân cách semicolon (;)
    if ";" in raw:
        lines = [ln.strip() for ln in raw.split(";") if ln.strip()]
        if lines:
            return lines
    return [raw]


def resolve_queries(args: argparse.Namespace) -> list[str]:
    """Xác định danh sách câu hỏi kiểm thử theo thứ tự ưu tiên:
    1. Cờ dòng lệnh --queries (file câu hỏi riêng biệt)
    2. Biến CHAT_QUERIES_FILE hoặc QUERIES_FILE trong .env
    3. Biến CHAT_QUERIES hoặc QUERIES trong .env (phân tách bởi |, xuống dòng, hoặc JSON array)
    4. Danh sách câu hỏi mẫu mặc định DEFAULT_QUERIES
    """
    env_file = getattr(args, "env_file", ".env")
    load_dotenv(env_file)

    # 1. Tham số dòng lệnh --queries
    if getattr(args, "queries", None):
        p = Path(args.queries)
        if not p.is_file():
            raise FileNotFoundError(f"Không tìm thấy file --queries: {args.queries}")
        lines = [ln.strip() for ln in p.read_text("utf-8").splitlines() if ln.strip()]
        if not lines:
            raise ValueError(f"File --queries rỗng: {args.queries}")
        print(f"queries  : {len(lines)} câu hỏi nạp từ cờ --queries ({p.name})")
        return lines

    # 2. File được trỏ tới từ biến môi trường CHAT_QUERIES_FILE / QUERIES_FILE
    queries_file = os.environ.get("CHAT_QUERIES_FILE") or os.environ.get("QUERIES_FILE")
    if queries_file:
        candidates = [
            Path(queries_file),
            Path.cwd() / queries_file,
            Path(__file__).resolve().parents[2] / queries_file,
        ]
        target = next((c for c in candidates if c.is_file()), None)
        if target:
            lines = [ln.strip() for ln in target.read_text("utf-8").splitlines() if ln.strip()]
            if lines:
                print(f"queries  : {len(lines)} câu hỏi nạp từ file {target.name} (qua .env)")
                return lines

    # 3. Chuỗi danh sách câu hỏi trực tiếp trong CHAT_QUERIES / QUERIES
    raw_env = os.environ.get("CHAT_QUERIES") or os.environ.get("QUERIES")
    if raw_env:
        parsed = parse_queries(raw_env)
        if parsed:
            print(f"queries  : {len(parsed)} câu hỏi nạp từ .env (CHAT_QUERIES)")
            return parsed

    # 4. Dự phòng mặc định
    print("queries  : dùng danh sách câu hỏi mẫu mặc định (chưa cấu hình CHAT_QUERIES trong .env)")
    return DEFAULT_QUERIES


def _mint_jwt(secret: str, ttl_s: int = 7200) -> str:
    """Tạo JWT nội bộ (server-to-server) để gửi trong header Authorization Bearer."""
    import jwt

    return jwt.encode(
        {"roles": ["super_admin"], "exp": int(time.time()) + ttl_s},
        secret,
        algorithm="HS256",
    )


class UserPool:
    """Quản lý danh sách người dùng và token xoay vòng theo từng request."""

    def __init__(self, users: list[tuple[str, str]]) -> None:
        self._users = users
        self._i = 0  # Con trỏ xoay vòng

    def __len__(self) -> int:
        return len(self._users)

    def next(self) -> tuple[str, str]:
        """Lấy thông tin (user_id, token) tiếp theo theo cơ chế xoay vòng (round-robin)."""
        user = self._users[self._i % len(self._users)]
        self._i += 1
        return user


def load_users(path: Path) -> list[tuple[str, str]]:
    """Đọc file CSV danh sách người dùng và trả về danh sách tuple (user_id, token).

    Hỗ trợ linh hoạt các tên cột phổ biến (không phân biệt chữ hoa/thường):
      - Cột token: 'token', 'user_token', 'access_token', 'api_key'
      - Cột user id: 'user_id', 'userid', 'id', 'user'
    """
    with path.open("r", encoding="utf-8", newline="") as f:
        reader = csv.DictReader(f)
        if not reader.fieldnames:
            raise ValueError(f"{path}: file rỗng, không có header")
        cols = {(name or "").strip().lower(): name for name in reader.fieldnames}

        # Tự động nhận diện cột token và user_id
        token_key = next(
            (k for k in ("token", "user_token", "access_token", "api_key") if k in cols),
            None,
        )
        if not token_key:
            raise ValueError(
                f"{path}: không tìm thấy cột chứa token. Yêu cầu có cột 'token' hoặc 'user_token'."
            )
        token_col = cols[token_key]

        user_key = next((k for k in ("user_id", "userid", "id", "user") if k in cols), None)
        user_col = cols[user_key] if user_key else None

        users: list[tuple[str, str]] = []
        for idx, row in enumerate(reader, 1):
            token = (row.get(token_col) or "").strip()
            if not token:
                continue
            uid = (row.get(user_col) or str(idx)).strip() if user_col else str(idx)
            users.append((uid, token))
    if not users:
        raise ValueError(f"{path}: không có dòng nào hợp lệ (mọi dòng đều thiếu token)")
    return users


def _payload(query: str, user_token: str) -> dict[str, Any]:
    """Tạo payload JSON gửi đến endpoint /chat mô phỏng một phiên trò chuyện thực tế."""
    return {
        "user_token": user_token,
        "summary": "",
        "history": [
            {"role": "user", "text": "Xin chào"},
            {
                "role": "assistant",
                "text": "Chào bạn, mình là chatbot. Bạn cần hỗ trợ gì?",
            },
            {"role": "user", "text": query},
        ],
    }


class ChatSSEScenario(Scenario):
    """Kịch bản kiểm thử API Chatbot hỗ trợ streaming SSE (endpoint POST /chat)."""

    name = "chat-sse"
    metric_label = "ttft"

    def __init__(
        self,
        url: str,
        headers: dict[str, str],
        timeout: float,
        pool: UserPool,
        queries: list[str],
        cache_bust: bool = False,
    ) -> None:
        self.url = url
        self.headers = headers
        self.timeout = timeout
        self.pool = pool
        self.queries = queries
        self.cache_bust = cache_bust

    @classmethod
    def from_args(
        cls, args: argparse.Namespace, headers: dict[str, str]
    ) -> ChatSSEScenario:
        """Khởi tạo ChatSSEScenario từ các tham số dòng lệnh."""
        hdrs = dict(headers)
        if "Content-Type" not in hdrs:
            hdrs["Content-Type"] = "application/json"

        token = args.jwt or (_mint_jwt(args.jwt_secret) if args.jwt_secret else "")
        if token and "Authorization" not in hdrs:
            hdrs["Authorization"] = f"Bearer {token}"

        queries = resolve_queries(args)

        # Xác định file users_csv (truyền qua cờ hoặc tự động nạp users.csv nếu tồn tại)
        users_csv: Path | None = None
        if getattr(args, "users_csv", None):
            p = Path(args.users_csv)
            if p.is_file():
                users_csv = p
            else:
                for cand in (
                    Path.cwd() / p,
                    Path(__file__).resolve().parents[2] / p,
                    Path(__file__).resolve().parents[3] / p,
                    Path(__file__).resolve().parent.parent / p,
                ):
                    if cand.is_file():
                        users_csv = cand
                        break
                if not users_csv:
                    print(
                        f"CẢNH BÁO: không thấy {args.users_csv}, dùng --user-token.",
                        file=sys.stderr,
                    )
        else:
            def_csv = find_default_users_csv()
            if def_csv.is_file():
                users_csv = def_csv

        if users_csv and users_csv.is_file():
            pool = UserPool(load_users(users_csv))
            print(f"pool     : {len(pool)} user từ {users_csv.name}")
        else:
            pool = UserPool([("", args.user_token)])
            print("pool     : 1 token dùng chung (--user-token)")

        # Hàng rào kiểm tra Rate-limit của tài khoản: len(pool) / max(steps) < 120s
        max_step = 0.0
        steps_str = getattr(args, "steps", "")
        if steps_str:
            try:
                max_step = max(float(s) for s in steps_str.split(",") if s.strip())
            except ValueError:
                pass

        if max_step > 0:
            cycle_s = len(pool) / max_step
            if cycle_s < 120.0:
                if len(pool) == 1:
                    print(
                        f"CẢNH BÁO: Đang dùng 1 token duy nhất cho đỉnh tải {max_step:g} req/s. "
                        "Nguy cơ chạm rate-limit của tài khoản thay vì đo công suất hệ thống.",
                        file=sys.stderr,
                    )
                else:
                    print(
                        f"CẢNH BÁO: Pool chỉ có {len(pool)} user cho đỉnh tải {max_step:g} req/s "
                        f"(chu kỳ lặp lại {cycle_s:.1f}s < 120s). "
                        "Nguy cơ chạm rate-limit của tài khoản thay vì đo công suất hệ thống.",
                        file=sys.stderr,
                    )

        return cls(
            url=args.url,
            headers=hdrs,
            timeout=args.timeout,
            pool=pool,
            queries=queries,
            cache_bust=args.cache_bust,
        )

    async def execute(self, client: httpx.AsyncClient) -> dict[str, Any]:
        """Gửi 1 request streaming SSE tới endpoint /chat và đo TTFT."""
        q = random.choice(self.queries)
        if self.cache_bust:
            q = f"{q} (ma tham chieu {random.randint(100000, 999999)})"
        uid, token = self.pool.next()
        body = _payload(q, token)

        t0 = time.perf_counter()
        rec: dict[str, Any] = {
            "type": "req",
            "t_start": time.time(),
            "user_id": uid,
            "ttft_ms": None,
            "notice_ms": None,
            "total_ms": None,
            "status": None,
            "ok": False,
            "err": None,
            "end_seen": False,
        }
        try:
            async with client.stream(
                "POST", self.url, json=body, headers=self.headers, timeout=self.timeout
            ) as resp:
                rec["status"] = resp.status_code
                if resp.status_code != 200:
                    await resp.aread()
                    rec["err"] = f"http_{resp.status_code}"
                    rec["total_ms"] = (time.perf_counter() - t0) * 1000
                    return rec

                event = ""
                async for line in resp.aiter_lines():
                    if line.startswith("event:"):
                        event = line[6:].strip()
                    elif line.startswith("data:"):
                        now = (time.perf_counter() - t0) * 1000
                        if event == "message" and rec["ttft_ms"] is None:
                            rec["ttft_ms"] = now
                        elif event == "notice" and rec["notice_ms"] is None:
                            rec["notice_ms"] = now
                        elif event == "message_end":
                            rec["end_seen"] = True
                        elif event == "error":
                            rec["err"] = "sse_error"

                rec["ok"] = rec["end_seen"] and rec["err"] is None
        except Exception as exc:
            rec["err"] = type(exc).__name__
        rec["total_ms"] = (time.perf_counter() - t0) * 1000
        return rec
